HB/common/phant-abbreviation.elpi

```diff
@@ -180,12 +183,23 @@ build-abbreviation K F [infer-type N sortclass|AL] K'' (fun N _ AbbrevFx) :- !,
 build-abbreviation K F [infer-type N funclass|AL] K'' (fun N _ AbbrevFx) :- !,
   pi x\ build-abbreviation K {mk-app F [{{ lib:hb.Phantom (_ -> _) lp:x }}]} AL K' (AbbrevFx x),
   K'' is K' + 1.
-build-abbreviation _ _ [infer-type _ (grefclass _)|_] _ _ :- coq.error "TODO".
+build-abbreviation K F [infer-type N (grefclass Class)|AL] K'' (fun N _ AbbrevFx) :- !,
+  build-type-pattern Class Pat,
+  pi x\ build-abbreviation K {mk-app F [{{ lib:hb.Phantom lp:Pat lp:x }}]} AL K' (AbbrevFx x),
+  K'' is K' + 1.
 build-abbreviation K F [implicit|AL] K' FAbbrev :- !,
   build-abbreviation K {mk-app F [_]} AL K' FAbbrev.
 build-abbreviation K F [unify|AL] K' FAbbrev :- !,
   build-abbreviation K {mk-app F [{{lib:@hb.id _ _}}]} AL K' FAbbrev.
 
+pred build-type-pattern i:gref, o:term.
+build-type-pattern GR Pat :- coq.env.typeof GR T,
```

Comment:
```suggestion
build-type-pattern GR Pat :-
```

Reply:
I think I'd like a comment for what this piece of code is supposed to do.
Isn't it something like
```
coq.safe-dest-app Ty (global GR) Args,
coq.subst-prod T Args {{ Type }}
```
?

Reply:
Indeed, comment added. The main idea is as follows:
* in case `sortclass`, we know the sort is of some type Type, so we unify with `Type`
* in case `funclass`, we know the sort is a function, so we unify with `_ -> _`
* in case `grefclass GR`, we know the sort is of type `GR x1 ... xn` when `GR : forall x1 ... xn, Type`, so we unify with `GR _ ... _` (with `n` holes)